src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilderCache.cs

```diff
@@ -32,14 +38,7 @@ public List<List<SelectionCriteria>> GetSelectionCriteria(RestoreTargetGraph gra
         {
             // Criteria are unique on graph and framework override.
             var key = new CriteriaKey(graph.TargetGraphName, framework);
-
-            if (!_criteriaSets.TryGetValue(key, out var criteria))
-            {
-                criteria = LockFileUtils.CreateOrderedCriteriaSets(graph, framework);
-                _criteriaSets.Add(key, criteria);
-            }
-
-            return criteria;
+            return _criteriaSets.GetOrAdd(key, _ => LockFileUtils.CreateOrderedCriteriaSets(graph, framework));
```

Comment:
> The difficulty with is that CreateOrderedCriteriaSets uses ManagedCodeConventions.ManagedCodeCriteria.ForFrameworkAndRuntime() and RuntimeIdentifier to create the criteria, but ManagedCodeCriteria is not equatable so it cannot be used as a member of CriteriaKey. On the other hand instances of ManagedCodeCriteria differ only by the runtime graph so the CriteriaKey is rather correct as it already does use runtime identifier as a part of the key..

Yeah, that's my reasoning as well! This a controlled codepath, so as long as we test it well, we should be good.

Reply:
So this one tricky, lol.

As of right now, ManagedCodeConventions *cannot* be easily shared. 

ManagedCodeConventions contains the runtime graph, which is provided by the SDK through `RuntimeIdentifierGraphPath` in .NET 3.x or later or a package in .NET 2.x and earlier.

https://github.com/NuGet/NuGet.Client/blob/50e383fa7cd94f9d5c8b0120ab415bf04a74ab4f/src/NuGet.Core/NuGet.Packaging/ContentModel/ManagedCodeConventions.cs#L106-L124

Given that *global.json* exists, we'd need to account for that. 
It's unfortunate that the differences in the runtime graph are not likely to cause problems often. It's almost seems like a theoretical problem only. 
We'd need to have 2 projects, with the same RID, getting the same package, but for *whatever* reason, one has a runtime graph with a runtime priority different from the other one. 

It's kind of difficult to follow all the inputs to a RuntimeGraph, because it also contains a list of runtimes specified by the project, but that doesn't affect the result of `AreCompatible` that I linked. 

The merging is happening here:
https://github.com/NuGet/NuGet.Client/blob/50e383fa7cd94f9d5c8b0120ab415bf04a74ab4f/src/NuGet.Core/NuGet.Commands/RestoreCommand/ProjectRestoreCommand.cs#L140-L144

The runtimegraph doesn't really carry any details about what files contain the runtime graph in the first place, either.

So 3 options: 

* Start by *not* caching when we're dealing with a RID specific graph. This would reduce the benefits of this change. Create a future issue that someone can pick up (might be you, if you choose so).
* Try to fix this issue by figuring out how to reason about RuntimeGraph equality for the purposes of ManagedCodeConventions. If we were to do this, I'd say we merge this change in a feature branch. The code here is pretty good, and I don't see the point in re-reviewing the same thing over and over again. 
* Assume RID graphs never conflict and just merge it.

Reply:
`RestoreTargetGraph.RuntimeGraph` already implements the `IEquatable` interface so to be 100% correct we could use it in the `CriteriaKey`? I'll measure whether it makes it much slower.

Reply:
@nkolev92 Hi, I've decided to go for option 1. To make reviewing easier I've created separate PR for this: https://github.com/marcin-krystianc/NuGet.Client/pull/1

Reply:
Given that the selection criteria is the thing that can be different, do we need an equivalent check here?

Reply:
Thinking about it more, caching the selection criteria might be unnecessary if we just pass it as a parameter to the method that calls it at https://github.com/NuGet/NuGet.Client/blame/d2e800b88d6b941732272c3a8cf7115836e98e22/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/LockFileUtils.cs#L74.

Reply:
Note my focus is on understand what the correct behavior should be. 

In the interest of getting this completed, I'd be happy to make the change at a later point. (or you are free to do it)

Reply:
I think it is ok as it is now.
- `ManagedCodeCriteria.ForFrameworkAndRuntime` doesn't really use `RuntimeGraph` so it is ok to cache results of `CreateOrderedCriteriaSets` even if they underlying `RuntimeGraphs` are different.
-  To my understanding, the `RuntimeIdentifier_CompatibilityTest` method is always called with the correct `ManagedCodeConventions/RuntimeGraph` -> https://github.com/marcin-krystianc/NuGet.Client/blob/dev-20210303-RestoreCommandCache/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/LockFileUtils.cs#L235-L239
-  If the `SelectionCriteria` caching was wrong then the test case called `LockFileBuilderCache_DifferentRuntimeGraph_WillNotCache` would fail